Add tag --restore/-r flag to tag command

saschagrunert opened the pull request:
It is now possible to undo tag operations if the names history is able
to restore the latest tag of a image.

rhatdan commented on cmd/podman/tag.go:
```diff
@@ -29,15 +29,20 @@ var (
+	flags := tagCommand.Flags()
+	flags.BoolVarP(
+		&tagCommand.Undo, "undo", "u", false,
+		"Undo the latest tag operation and restore the previous tag",
```
Why only the latest tag?  How is this different the rm of a tag?
I would really like to see an explanation, example of what you are trying to do?

podman tag IMAGE foo
podman tag IMAGE bar
What do I have when I do
podman tag IMAGE --undo?

saschagrunert replied:
I think we cannot remove a tag right now in podman, right? I was thinking about to add another option to be able to remove a tag from a image.

Right now it works like (only) like this:
```bash
# Downloaded alpine, which has no history initially
> ./bin/podman images --history
REPOSITORY                 TAG      IMAGE ID       CREATED       SIZE      HISTORY
docker.io/library/alpine   latest   965ea09ff2eb   7 weeks ago   5.82 MB   <none>

# Retagging alpine to `test`, now both have a history
> ./bin/podman tag docker.io/library/alpine test
> ./bin/podman images --history
REPOSITORY                 TAG      IMAGE ID       CREATED       SIZE      HISTORY
docker.io/library/alpine   latest   965ea09ff2eb   7 weeks ago   5.82 MB   docker.io/library/alpine:latest
localhost/test             latest   965ea09ff2eb   7 weeks ago   5.82 MB   docker.io/library/alpine:latest

# Now we're able to restore the previous tag of `test` by
> ./bin/podman tag localhost/test --restore
> ./bin/podman images --history
REPOSITORY                 TAG      IMAGE ID       CREATED       SIZE      HISTORY
docker.io/library/alpine   latest   965ea09ff2eb   7 weeks ago   5.82 MB   localhost/test:latest
```
We probably can also undo the tag when specifying the `docker.io/library/alpine` image, but then I would have to add more logic. Would this be intended?

rhatdan replied:
```
podman tag alpine foobar
podman tag alpine alpine:foobar
podman tag alpine docker.io/library/alpine:foobar
```
```
podman tag alpine --restore
```
I take it this just removes just docker.io/library/alpine:foobar?
What happens if I did?

podman tag foobar --restore?

I might be bikes heading here but would it be better to have the syntax be.
podman tag --remove docker.io/library/alpine:foobar
And then it just removes the tag, and cleans up the history?

If I eventually do a 
podman tag --remove docker.io/library/alpine:latest
then it would generate a 
<none> <none> image?

rhatdan replied:
Another option would be to add a 
podman untag ...

vrothberg commented on libpod/image/image.go:
```diff
@@ -577,6 +577,30 @@ func (i *Image) TagImage(tag string) error {
+
+	history := i.NamesHistory()
+	if len(history) < 1 {
+		return errors.New("provided image has to tag history to undo")
```
Can we get the image name in the error message as well? This will make the logs more explicit (same below).

saschagrunert replied:
Yep, fixed :+1:

vrothberg commented on libpod/image/image.go:
```diff
@@ -577,6 +577,30 @@ func (i *Image) TagImage(tag string) error {
 	return nil
 }
 
+func (i *Image) UndoTag() error {
```
Please add a (verbose) comment.

saschagrunert replied:
Added a comment as suggested.

vrothberg commented on docs/source/markdown/podman-tag.1.md:
````diff
@@ -19,14 +19,19 @@ the *image* and the *target-name*.
 $ podman tag httpd myregistryhost:5000/fedora/httpd:v2
-```
 
+$ podman tag fedora:latest --undo
````
I'd use `0e3bbc2` to illustrate that the ID is static but the tags are somehow arbitrary.

saschagrunert replied:
Changed as suggested. :+1:

saschagrunert commented:
> Wow, you're on  today, @saschagrunert :)
> 
> Just a couple of nits. I like the change and kudos for supporting the remote-client as well!
> 
> Regarding name, I prefer calling it `--restore` as it's clearer (to me) that we restore the previous tag. But that's just a minor preference.

I addressed your review comments and renamed the option to `--restore`. I just have to check if the tests are happy.

mheon commented:
Agree with Dan, an untag command sounds better than an undo flag.

On Fri, Dec 13, 2019, 09:10 Daniel J Walsh <notifications@github.com> wrote:

> *@rhatdan* commented on this pull request.
> ------------------------------
>
> In cmd/podman/tag.go
> <https://github.com/containers/libpod/pull/4693#discussion_r357661586>:
>
> > @@ -29,15 +29,20 @@ var (
>
>  func init() {
>  	tagCommand.Command = _tagCommand
> +
> +	flags := tagCommand.Flags()
> +	flags.BoolVarP(
> +		&tagCommand.Undo, "undo", "u", false,
> +		"Undo the latest tag operation and restore the previous tag",
>
> Another option would be to add a
> podman untag ...
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/containers/libpod/pull/4693?email_source=notifications&email_token=AB3AOCBIDKP7JV4GDBBMEZLQYOJURA5CNFSM4J2L44L2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPEEB2Y#discussion_r357661586>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB3AOCEZLGMSV75SQ5CXO6LQYOJURANCNFSM4J2L44LQ>
> .
>

saschagrunert commented:
Not needed any more
